src/tests/Interop/COM/ComWrappers/API/Program.cs

```diff
@@ -962,6 +963,172 @@ public void ValidateAggregationWithReferenceTrackerObject()
+        {
+            protected override ComInterfaceEntry* ComputeVtables(object obj, CreateComInterfaceFlags flags, out int count)
+            {
+                count = 0;
+                return null;
+            }
+
+            protected override object CreateObject(IntPtr externalComObject, CreateObjectFlags flags)
+            {
+                return new object();
+            }
+
+            protected override void ReleaseObjects(IEnumerable objects)
+            {
+                throw new NotImplementedException();
+            }
+        }
+
+        [Fact]
+        [PlatformSpecific(TestPlatforms.Windows)] // COM apartments are Windows-specific
+        public unsafe void CrossApartmentQueryInterface_NoDeadlock()
+        {
+            Console.WriteLine($"Running {nameof(CrossApartmentQueryInterface_NoDeadlock)}...");
+            using ManualResetEvent hasAgileReference = new(false);
+            using ManualResetEvent testCompleted = new(false);
+
+            IntPtr agileReference = IntPtr.Zero;
+            try
+            {
+                Thread staThread = new(() =>
+                {
+                    var cw = new RecursiveSimpleComWrappers();
+                    IntPtr comObject = cw.GetOrCreateComInterfaceForObject(new Test(), CreateComInterfaceFlags.None);
+                    try
+                    {
+                        Marshal.ThrowExceptionForHR(RoGetAgileReference(0, IUnknownVtbl.IID_IUnknown, comObject, out agileReference));
+                    }
+                    finally
+                    {
+                        Marshal.Release(comObject);
+                    }
+                    hasAgileReference.Set();
+                    testCompleted.WaitOne();
+                });
+                staThread.SetApartmentState(ApartmentState.STA);
+
+                Thread mtaThread = new(() =>
+                {
+                    hasAgileReference.WaitOne();
+                    IntPtr comObject;
+                    int hr = ((delegate* unmanaged<IntPtr, in Guid, out IntPtr, int>)(*(*(void***)agileReference + 3 /* IAgileReference.Resolve slot */)))(agileReference, IUnknownVtbl.IID_IUnknown, out comObject);
+                    Marshal.ThrowExceptionForHR(hr);
+                    try
+                    {
+                        var cw = new RecursiveSimpleComWrappers();
+                        // Make sure that ComWrappers isn't locking in GetOrCreateObjectForComInstance
+                        // across the QI call
+                        // by forcing marshalling across COM apartments.
+                        _ = cw.GetOrCreateObjectForComInstance(comObject, CreateObjectFlags.None);
+                    }
+                    finally
+                    {
+                        Marshal.Release(comObject);
+                    }
+                    testCompleted.Set();
+                });
+                mtaThread.SetApartmentState(ApartmentState.STA);
```

Comment:
Did you mean `MTA` here? 😅
```suggestion
                mtaThread.SetApartmentState(ApartmentState.MTA);
```

Reply:
Yep... Did a few more fixes on the new test as well.